src/offline/client.cc

```diff
@@ -353,52 +354,97 @@ PostInstallAction install(const Config& cfg_in, const UpdateSrc& src) {
     throw std::runtime_error("Failed to download Target; err: " + download_res.description);
   }
 
+  PostInstallAction post_install_action{PostInstallAction::Undefined};
+  if (client->getCurrent().sha256Hash() != target.sha256Hash()) {
+    const auto install_res = client->install(target);
+    if (install_res != data::ResultCode::Numeric::kNeedCompletion) {
+      throw std::runtime_error("Failed to install Target");
+    }
+    post_install_action = PostInstallAction::NeedReboot;
+  } else {
+    // don't `install` since it will create/run containers and we don't want to do it
+    // before we register images and restart dockerd
+    client->storage->savePrimaryInstalledVersion(target, InstalledVersionUpdateMode::kPending);
+    post_install_action = PostInstallAction::NeedDockerRestart;
+  }
+
   const auto pacman_cfg{ComposeAppManager::Config(cfg_in.pacman)};
   registerApps(target, pacman_cfg.reset_apps_root, pacman_cfg.images_data_root);
 
-  if (client->getCurrent().sha256Hash() == target.sha256Hash()) {
-    // run apps after dockerd reload
-    return PostInstallAction::NeedDockerRestart;
-  };
-  const auto install_res = client->install(target);
-  if (install_res != data::ResultCode::Numeric::kNeedCompletion) {
-    throw std::runtime_error("Failed to install Target");
-  }
-
-  return PostInstallAction::NeedReboot;
+  return post_install_action;
 }
 
-void run(const Config& cfg_in, const UpdateSrc& src, std::shared_ptr<HttpInterface> docker_client_http_client) {
-  auto client{createOfflineClient(cfg_in, src, docker_client_http_client)};
+PostRunAction run(const Config& cfg_in, std::shared_ptr<HttpInterface> docker_client_http_client) {
+  auto client{createOfflineClient(cfg_in,
+                                  /* src dir is not needed in the case of run command */
+                                  {
+                                      "unknown-tuf-dir",
+                                      "unknown-ostree-dir",
+                                      "unknown-apps-dir",
+                                  },
```

Comment:
I don't like it neither. This is one of the consequences of having a notion of package manager instead of having separate entities for download, install and run and/or not combining their implementation for different types of targets in one place (compose app manager).
Anyway, initially, I tried another approach. As a matter of fact, "finalization" and/or Apps starting can be performed by not "tailored" LiteClient instance (only download requires all those dep injection and customization), hence no need in the src param at all. However, it requires more code and is less error prone. Some changes to LiteClient may lead to communication to cloud at this stage, having these dummy values will help to quickly identify such issues since they are supposed to appear in logs.